File: mqtt/mqttManager.py
import yaml
import paho.mqtt.client as mqttClient
from pathlib import Path
import re

class MqttStarter:
    __mosquittoClient = None
    __topicsToSubscribe = None
    __config = None
    __mosquittoHandlers = {}

    def __init__(self) -> None:
        self.__loadConfigFile()
        self.__mosquittoClient = mqttClient.Client(mqttClient.CallbackAPIVersion.VERSION1)
        self.__topicsToSubscribe = self.__getTopicsToSubscribe()

    def __loadConfigFile(self):
        try:
            self.__config = yaml.safe_load(open(str(Path.cwd()) + "/config.yaml"))
        except:
            raise Exception("Unable to read config file")
        # Validating the config file is still pending

    def __getTopicsToSubscribe(self) -> list[tuple[str, int]]:
        handlers = self.__config["mqtt-handlers"]
        result = []
        for itemkey, itemValue in handlers.items():
            self.__mosquittoHandlers[itemkey] = itemValue
            result.append((itemValue["topic"], itemValue["qos"]))
        return result
    
    @staticmethod
    def mqtt_wildcard_match(subscription_pattern, topic):
        pattern = re.escape(subscription_pattern)
        pattern = pattern.replace(r'\+', '[^/]+')
        pattern = pattern.replace(r'\#', '.*')
        return re.match(pattern + '$', topic) is not None

    def __onConnect(self):
        def on_connect(client, userdata, flags, rc):
            if rc == 0:
                print("Connected to broker")
                self.__mosquittoClient.subscribe(self.__topicsToSubscribe)
            else:
                print("Connection failed")
                raise Exception("Unable to connect to mosquitto server")
        return on_connect

    def __onMessage(self, messageHandler):
        def on_message(client, userdata, message):
            handler = next(((itemKey, itemValue) for itemKey, itemValue in self.__mosquittoHandlers.items() if self.mqtt_wildcard_match(itemValue["topic"], message.topic)), None)
            if (handler != None):
                messageHandler(handler, message)
        return on_message

    def ConnectToMosquitto(self, messageHandler):
        mosquittoConfig = self.__config["mosquitto"]
        broker_address= mosquittoConfig["host"]
        port = mosquittoConfig["port"]
        user = mosquittoConfig["user"]
        password = mosquittoConfig["pass"]
        self.__mosquittoClient.username_pw_set(user, password=password)
        self.__mosquittoClient.on_connect= self.__onConnect()
        self.__mosquittoClient.on_message= self.__onMessage(messageHandler)
        self.__mosquittoClient.connect(broker_address, port=port)
        self.__mosquittoClient.loop_start()

    def DisconnectMosquitto(self):
        self.__mosquittoClient.disconnect()
        self.__mosquittoClient.loop_stop()

    

File: src/mqtt/mqttManager.ts
import mqtt from 'mqtt';
import yaml from 'js-yaml';
import fs from 'fs';
import path from 'path';
import { OperationCanceledException } from 'typescript';

interface MosquittoConfig {
    user: string;
    pass: string;
    host: string;
    port: number;
}

interface HandlerConfig {
    topic: string;
    qos: number;
}

export class MqttStarter {
    private mosquittoClient!: mqtt.MqttClient;
    private config: any; // Ideally, define a more specific type
    private topicsToSubscribe!: Array<[string, number]>;
    private mosquittoHandlers: { [key: string]: HandlerConfig } = {};

    constructor() {
      this.config = this.loadConfigFile();
    }

    public ConnectToMosquitto(callBack: Function) {
      this.mosquittoClient = mqtt.connect(this.getMosquittoUrl(this.config.mosquitto));
      this.topicsToSubscribe = this.getTopicsToSubscribe();

      this.mosquittoClient.on('connect', () => {
          console.log("Connected to broker");
          this.topicsToSubscribe.forEach(topic => {
              this.mosquittoClient.subscribe(topic[0]);
          });
      });

      this.mosquittoClient.on('message', (topic, message) => {
          let handler = this.findHandlerForTopic(topic);
          if (handler) {
              this.handleMessage(handler, message);
          }
      });

      this.mosquittoClient.on('error', (error) => {
        console.error("Connection failed:", error);
        throw new OperationCanceledException()
      });
    }

    private loadConfigFile(): any { // Should define an interface for the return type
        try {
            const configPath = path.join(process.cwd(), 'config.yaml');
            const fileContents = fs.readFileSync(configPath, 'utf8');
            return yaml.load(fileContents);
        } catch (error) {
            throw new Error("Unable to read config file");
        }
    }

    private getMosquittoUrl(mosquittoConfig: MosquittoConfig): string {
        return `mqtt://${mosquittoConfig.user}:${mosquittoConfig.pass}@${mosquittoConfig.host}:${mosquittoConfig.port}`;
    }

    private getTopicsToSubscribe(): Array<[string, number]> {
        const handlers = this.config['mqtt-handlers'];
        let topics: Array<[string, number]> = [];
        for (const key in handlers) {
            this.mosquittoHandlers[key] = handlers[key];
            topics.push([handlers[key]['topic'], handlers[key]['qos']]);
        }
        return topics;
    }

    private findHandlerForTopic(topic: string): [string, HandlerConfig] | null {
        for (const key in this.mosquittoHandlers) {
            if (this.mqttWildcardMatch(this.mosquittoHandlers[key]['topic'], topic)) {
                return [key, this.mosquittoHandlers[key]];
            }
        }
        return null;
    }

    private mqttWildcardMatch(subscriptionPattern: string, topic: string): boolean {
        const pattern = subscriptionPattern.replace(/\+/g, '[^\/]+').replace(/#/g, '.*');
        return new RegExp(`^${pattern}$`).test(topic);
    }

    private handleMessage(handler: [string, HandlerConfig], message: any): void {
        console.log(`Handling message for ${handler[0]}: ${message.toString()}`);
    }

    public disconnect(): void {
        this.mosquittoClient.end();
    }
}
